ffi/orcjit.cpp

```diff
@@ -0,0 +1,337 @@
+
+inline TargetMachine *unwrap(LLVMTargetMachineRef TM) {
+    return reinterpret_cast<TargetMachine *>(TM);
+}
+
+inline LLVMOrcJITTargetMachineBuilderRef wrap(JITTargetMachineBuilder *JTMB) {
+    return reinterpret_cast<LLVMOrcJITTargetMachineBuilderRef>(JTMB);
+}
+
+class JITDylibTracker {
+  public:
+    std::shared_ptr<LLJIT> lljit;
+    IntrusiveRefCntPtr<llvm::orc::ResourceTracker> tracker;
+    JITDylib &dylib;
+    JITDylibTracker(std::shared_ptr<LLJIT> &lljit_, JITDylib &dylib_,
+                    IntrusiveRefCntPtr<llvm::orc::ResourceTracker> &&tracker_)
+        : lljit(lljit_), dylib(dylib_), tracker(tracker_) {}
+};
+
+typedef struct {
+    uint8_t element_kind;
+    char *value;
+    size_t value_len;
+} LinkElement;
+typedef struct {
+    char *name;
+    uint64_t address;
+} SymbolAddress;
+extern "C" {
+
+API_EXPORT(std::shared_ptr<LLJIT> *)
+LLVMPY_CreateLLJITCompiler(LLVMTargetMachineRef tm, const char **OutError) {
+    LLJITBuilder builder;
+
+    if (tm) {
+        // The following is based on
+        // LLVMOrcJITTargetMachineBuilderCreateFromTargetMachine. However,
+        // we can't use that directly because it destroys the target
+        // machine, but we need to keep it alive because it is referenced by
+        // / shared with other objects on the Python side.
+        auto *template_tm = unwrap(tm);
+
+        builder.setJITTargetMachineBuilder(
+            JITTargetMachineBuilder(template_tm->getTargetTriple())
+                .setCPU(template_tm->getTargetCPU().str())
+                .setRelocationModel(template_tm->getRelocationModel())
+                .setCodeModel(template_tm->getCodeModel())
+                .setCodeGenOptLevel(template_tm->getOptLevel())
+                .setFeatures(template_tm->getTargetFeatureString())
+                .setOptions(template_tm->Options));
+    }
+    builder.setObjectLinkingLayerCreator(
+        [&](llvm::orc::ExecutionSession &session, const llvm::Triple &triple)
+            -> std::unique_ptr<llvm::orc::ObjectLayer> {
+            auto linkingLayer =
+                std::make_unique<llvm::orc::RTDyldObjectLinkingLayer>(
+                    session, []() {
+                        return std::make_unique<llvm::SectionMemoryManager>();
+                    });
+            if (triple.isOSBinFormatCOFF()) {
+                linkingLayer->setOverrideObjectFlagsWithResponsibilityFlags(
+                    true);
+                linkingLayer->setAutoClaimResponsibilityForObjectSymbols(true);
+            }
+            linkingLayer->registerJITEventListener(
+                *llvm::JITEventListener::createGDBRegistrationListener());
+            return linkingLayer;
+        });
+
+    auto jit = builder.create();
+
+    if (!jit) {
+        char *message = LLVMGetErrorMessage(wrap(jit.takeError()));
+        *OutError = LLVMPY_CreateString(message);
+        LLVMDisposeErrorMessage(message);
+        return nullptr;
```

Comment:
Can we get to `llvm::orc::ExecutionSession::setErrorReporter` and define a reporter that doesn't print to stderr?

Reply:
done in https://github.com/numba/llvmlite/pull/942/commits/d7a0d6c1759a6dce021b1332afcfa62565bdce4a